packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -759,49 +755,302 @@ class CupertinoActionSheetAction extends StatelessWidget {
+  }
+}
+
+// The keys used by the dividers in the action section of an action sheet.
+//
+// The buttons and the dividers are placed alternately, both keyed by indexes.
+// Dividers use this special class so that the two lists of indexes can be told
+// apart.
+class _DividerKey extends ValueKey<int> {
+  const _DividerKey(super.value);
+}
+
+typedef _PressedUpdateHandler = void Function(int actionIndex, bool state);
+
+// The list of actions in an action sheet.
+//
+// This excludes the divider between the action section and the content section.
+class _ActionSheetActionSection extends StatelessWidget {
+  const _ActionSheetActionSection({
+    required this.actions,
+    required this.pressedIndex,
+    required this.dividerColor,
+    required this.backgroundColor,
+    required this.onPressedUpdate,
+    required this.scrollController,
+  });
+
+  final List<Widget>? actions;
+  final _PressedUpdateHandler onPressedUpdate;
+  final int? pressedIndex;
+  final Color dividerColor;
+  final Color backgroundColor;
+  final ScrollController? scrollController;
+
+  @override
+  Widget build(BuildContext context) {
+    if (actions == null || actions!.isEmpty) {
+      return const LimitedBox(
+        maxWidth: 0,
+        child: SizedBox(width: double.infinity, height: 0),
+      );
+    }
+    final List<Widget> column = <Widget>[];
+    for (int actionIndex = 0; actionIndex < actions!.length; actionIndex += 1) {
+      if (actionIndex != 0) {
+        column.add(_ActionSheetDivider(
+          key: _DividerKey(actionIndex),
+          dividerColor: dividerColor,
+          hidden: pressedIndex == actionIndex - 1 || pressedIndex == actionIndex,
+        ));
+      }
+      column.add(_ActionSheetButtonBackground(
+        key: ValueKey<int>(actionIndex),
+        onStateChange: (bool state) {
+          onPressedUpdate(actionIndex, state);
+        },
+        child: actions![actionIndex],
+      ));
+    }
+    return CupertinoScrollbar(
+      controller: scrollController,
+      child: SingleChildScrollView(
+        controller: scrollController,
+        child: Column(
+          children: column,
+        ),
+      ),
+    );
+  }
+}
+
+// The part of an action sheet without the cancel button.
+class _ActionSheetMainSheet extends StatefulWidget {
+  const _ActionSheetMainSheet({
+    required this.scrollController,
+    required this.actions,
+    required this.hasContent,
+    required this.contentSection,
+    required this.dividerColor,
+  });
+
+  final ScrollController? scrollController;
+  final List<Widget>? actions;
+  final bool hasContent;
+  final Widget contentSection;
+  final Color dividerColor;
+
+  @override
+  _ActionSheetMainSheetState createState() => _ActionSheetMainSheetState();
+}
+
+class _ActionSheetMainSheetState extends State<_ActionSheetMainSheet> {
+  int? _pressedIndex;
+  double _topOverscroll = 0;
+  double _bottomOverscroll = 0;
+
+  // Fills the overscroll area at the top and bottom of the sheet. This is
```

Comment:
If I understand correctly, this is because some pressed down button will have a different color so this code let's the button paint its background instead of this top level container.

Can we instead paint the basic background in _ActionSheetMainSheet and the pressed button to paint its own color on top?

Reply:
also should watch out for one frame delay if you are changing the size based on scroll notification.

Reply:
> Can we instead paint the basic background in _ActionSheetMainSheet and the pressed button to paint its own color on top?

I thought about it, but I'm not sure if it will be exactly as good, because both the basic background and the pressed background are translucent. 

> also should watch out for one frame delay if you are changing the size based on scroll notification.

Are you certain there will be a one frame delay, and the delay will cause a overlap and a gap on the overscroll (if that's the concern)?

I don't think that will happen, because to my understanding, after the pointer event, the gesture recognizer update the position of the scroll view with a `setState` and immediately uses the scroll notification to update the overscrolls' sizes. Therefore both change should take place at the same frame, the next frame.

I've also updated the test `Actions section correctly renders overscrolls` to verify the rendering result at the _next frame_ of the gesture, and no overlaps or gaps can be observed.

![image](https://github.com/flutter/flutter/assets/1596656/d84478ed-c168-4cf2-b93d-a8570aeed0de)

```dart
    await gesture.moveBy(const Offset(0, -300));
    // Test the bottom overscroll. Use `pump` not `pumpAndSettle` to verify the
    // rendering result of the immediate next frame.
    await tester.pump();
    await expectLater(
      find.byType(CupertinoActionSheet),
      matchesGoldenFile('cupertinoActionSheet.overscroll.2.png'),
    );
```

Reply:
> Are you certain there will be a one frame delay, and the delay will cause a overlap and a gap on the overscroll (if that's the concern)?

So I looks more into the overscroll notification code, it looks like the call loop will call into to the update method when as soon as user drag. so I think there shouldn't be a delay. sorry for the confusion.

> I thought about it, but I'm not sure if it will be exactly as good, because both the basic background and the pressed background are translucent. We'll have to simulate one translucent color by combining two translucent colors.

My original thought was trying to remove this weird additional padding, needed when overscroll. but I do agree it may be hard to find the right color to paint on top if they are translucent. The only potentional problem I can think of now since there won't be a frame delay will then be possible anti-aliasing line in the gap. but I don't think that is something we need to worry about not. 

LGTM for this approach then